core/src/main/java/org/apache/iceberg/ManifestReader.java

```diff
@@ -369,4 +390,28 @@ static List<String> withStatsColumns(Collection<String> columns) {
       return projectColumns;
     }
   }
+
+  private static <F extends ContentFile<F>> Function<ManifestEntry<F>, ManifestEntry<F>> idAssigner(
+      Long firstRowId) {
+    if (firstRowId != null) {
```

Comment:
Just checking, this is protecting against delete manifests and data manifests written before first-row-id is assignable?

Reply:
It ends up being multiple things. One case is when the manifest itself doesn't have a `first_row_id` because it was in an upgraded snapshot. Another is when this class is reading a delete manifest.